### DISCUSSION TITLE
How to best work with custom `graphene.ObjectType`s with the optimizer?

### DISCUSSION THREAD

**Discussion by rsomani95** [MrThearMan/graphene-django-query-optimizer#63] [Q&A] [answered]
In my project, I've been using a customised version of the [`django-taggit`](https://github.com/jazzband/django-taggit) library.

Before diving into specifics, at a high level, this is how it works:
Any django model can be "taggable" (i.e. have tags associated with them). These tags are represented in 2 tables -- one that keeps track of the unique tag names itself, and the other that keeps track of which taggable objects are associated with these tags.

Here's what it looks like in code:

First, the "taggable" code:

```python

from taggit.models import GenericUUIDTaggedItemBase, TagBase
from taggit.managers import TaggableManager

class Tag(TagBase):
    class Meta:
        unique_together = ("name", "category")

    category = models.CharField(
        db_index=True, default="custom", max_length=200, null=False
    )
    name = models.CharField(db_index=True, max_length=255)


class TaggedItem(GenericUUIDTaggedItemBase):
    class Meta:
        unique_together = [["content_type", "object_id", "tag"]]

    confidence = models.FloatField()


class Taggable(models.Model):
    class Meta:
        abstract = True

    tags = TaggableManager(blank=True, through=TaggedItem)
    tagged_through = TaggedItem

```

Then, any model that needs to be tagged can inherit the `Taggable` object, like so:

```python
class CustomModel(Taggable):
    ....
```

---

On the GQL side, I represent this as custom object type, which is defined as follows:

```python
import graphene

class OzuTag(graphene.ObjectType):
    category = graphene.String()
    library_id = graphene.String()
    name = graphene.String()
    user_id = graphene.String()


class OzuTaggedItem(OzuTag):
    confidence = graphene.Float()
    tag_type = graphene.String()

class CustomObjectNode((DjangoObjectType):
    tags = graphene.List(OzuTaggedItem)

    # The original code has more robustness checks and if-else clauses, but I've reproduced
    # a minimal version here for ease of demonstration
    def resolve_tags(self, info):
        res = []
        for tagged_item in self.prefetched_tagged_items:
            res.append(
                {
                    "name": tagged_item.tag.name,
                    "category": tagged_item.tag.category,
                    "user_id": tagged_item.tag.user_id,
                    "library_id": tagged_item.tag.library_id,
                    "confidence": tagged_item.confidence,
                    "tag_type": tagged_item.tag_type,
                }
            )


class Query:
    all_objects = DjangoFilterConnectionField(CustomObjectNode)

    def resolve_all_objects(self, info):
        qset = CustomObject.objects.all()
        is_querying_for_tags, _ = check_for_field(info, "tags")  # Checks incoming GQL query to see if 'tags' is queried for
        if is_querying_for_tags:
            tagged_items_prefetch = Prefetch(
                "tagged_items",
                queryset=(
                    TaggedItem.objects.filter(
                        object_id__in=qset.values_list("pk")
                    ).select_related("tag")
                ),
                to_attr="prefetched_tagged_items",
            )
            qset = qset.prefetch_related(tagged_items_prefetch)
        return qset
```

The above code is inefficient in that it doesn't do `.only()` optimization and feels hacky as I'm introspecting the `info` manually. However, I'm not sure how to best use this library to do the above. I'm unable to use the above logic with this library because I run into `"Unhandled field options type: <ObjectTypeOptions name='OzuTaggedItem'>",` when querying for `tags` in the GQL console. It's unclear to me where/how to tell the optimizer that `OzuTaggedItem` is a custom object type that has custom logic, and if this is currently supported? If I can utilize the library to do the prefetch optimizations directly, then even better! (I imagine this is possible given the source code I've scoured, just not sure what the right mechanism is), but if not, it'd be super useful to be able to override the optimizer logic for the `tags` field somehow. 

Also want to mention that if this isn't supported but can be, I'd be happy to put in a PR with some guidance.
Thanks!

**Comment by MrThearMan** [comment]
The issue here is that `tags` inherits from `graphene.ObjectType` and not `graphene_django.DjangoObjectType`. The optimizer only supports optimizing the latter (and connection types containing them), since it needs to know some information about the model the ObjectType is assosiated with to do the optimization, which you need to give in it in the `Meta` class.

I might consider adding an option to hook into the optimization process with settings, but currently #61 and #62 take precedence. After those are resolved (I'm working on them), I'm willing to accept solutions for this.

**Reply by rsomani95** [reply]
@MrThearMan I'm preparing a writeup for why inheriting `graphene_django.DjangoObjectType` doesn't do the optimal thing either (although I am currently working off the `0.1.11` release). In the meantime, do you think it's a good idea to allow certain fields to be skipped by the optimizer?

In this case, `django-taggit` does a lot of optimization behind the scenes, and given its quirky model relationships, it would be really useful to have the ability to let the optimizer ignore this field and handle it manually

**Comment by rsomani95** [comment]
Closing this discussion as there's now an ongoing PR for this: https://github.com/MrThearMan/graphene-django-query-optimizer/pull/67

**Comment by MrThearMan** [comment, marked as answer]
For closure, this has been solved in [v0.5.1](https://github.com/MrThearMan/graphene-django-query-optimizer/releases/tag/v0.5.1).

Since I also [mentioned adding hooks](https://github.com/MrThearMan/graphene-django-query-optimizer/discussions/63#discussioncomment-8588303) for configuring parts of the optimizer in this discussion, I'm announcing here that this is now possible in [v0.6.0](https://github.com/MrThearMan/graphene-django-query-optimizer/releases/tag/v0.6.0). Users can now simply subclass `OptimizationCompiler`, `QueryOptimizer`, `FilterInfoCompiler`, or `FieldSelectionCompiler`, and the library will then use those subclasses internally. This allows modifying specific parts of them without rewriting large parts of the library that use those classes.
